### ISSUE TITLE
systemd-resolved does not parse comments in /etc/hosts correctly

### ISSUE THREAD

**Issue by madumlao** [systemd/systemd#10779]
## Previous title: systemd-resolved should not cache /etc/hosts entries

(EDIT: actual issue found in https://github.com/systemd/systemd/issues/10779#issuecomment-438958883)

**Is your feature request related to a problem? Please describe.**
I am a web developer / system administrator that does extensive local testing of various environments before enabling them to production.

To facilitate such testing, I often add or change an `/etc/hosts` entry to point the domain to the system under test so that I can locally confirm that changes have been effective before making a change to the domain name. For example, I may point `foo.com` to `1.2.3.4` temporarily so that I can confirm the server on `1.2.3.4` matches the behavior wanted in production, and only when I can do so will I update the public DNS entry of `foo.com` to `1.2.3.4`.

I recently upgraded my system from Ubuntu 1604 (which ships with dnsmasq as a local resolver) to Ubuntu 1804 (which includes systemd-resolved) (systemd-237). After the switch I noticed a change in the system behavior I was unused to: even after removing a `/etc/hosts` entry from the file, systemd-resolved will still resolve the old `/etc/hosts` entries as valid answers. This is unexpected and completely disrupts the testing workflow. It also only affects systemd-resolved: upon restoring dnsmasq, I can confirm that queries to only dnsmasq (`dig foo @127.0.0.1`) do not show the deleted hosts entries, whereas queries to systemd-resolved (`dig foo @127.0.0.53`) do.

I cannot seem to get systemd-resolved to drop my old `/etc/hosts` entries by flushing its caches with `systemd-resolve --flush-caches` or `sudo kilall -USR2 systemd-resolved` or disabling caches under `/etc/systemd/resolved.conf`. However, even if I could, none of these are ideal: I _want_ a local DNS cache (for faster Internet), but I _don't_ want the cache to apply to `/etc/hosts`, which itself is a manually maintained cache / developer tool to begin with. As such I have reverted my system to dnsmasq for the moment.

**Describe the solution you'd like**
systemd-resolved entries that were resolved from `/etc/hosts` should not be cached at all, and should _always_ result in a hit to `/etc/hosts`. This way, adminstrators and developers that rely on /etc/hosts in their test workflow will be able to change such entries and immediately observe their changes in their live systems.

**Describe alternatives you've considered**
systemd-resolved _may_ cache `/etc/hosts` entries, but should give a very low validity, on the order of a few seconds, if this is needed to prevent disk hits. However, I would argue that the kernel disk cache should be more than sufficient for `/etc/hosts` anyways.

**Comment by madumlao** [comment]
I would also like to add that upon reverting my system to dnsmasq, I ran into an issue with the combination of systemd-resolved, resolvconf, and dnsmasq. This ate up roughly 3 hours of my time figuring out what was changing what, but out of the box, disabling systemd-resolved and enabling dnsmasq breaks your system.

The long and short of it is that the default setup of systemd installs an entry inside of `/run/resolvconf/interface` that points DNS entries to `127.0.0.53`, or the default systemd-resolved address.

The problem is that DNS forwarder software like dnsmasq will also pick up local entries from `/run/resolvconf/interface` due to a configuration in `resolvconf/update.d`. DNS forwarder software like dnsmasq wants to read `/run/resolvconf/interface` entries to see if e.g. networkmanager, dhclient, or other software added nameservers that it will for its own resolv.conf file. But since systemd-resolved puts an entry there for `127.0.0.53`, all other DNS forwarder software that does the default behavior will be forwarded to systemd-resolved, even if systemd-resolved is not on or active. i.e., if you naively disable systemd-resolved, your system will just break.

Would you happen to know what puts an entry in `/run/resolvconf/interface/systemd-resolved` (startup script? systemd-resolved default behavior? dhclient? network-manager?) and how I might disable that?  If it's provided by systemd-resolved itself, could we have a switch to disable it, or at least by default, remove the resolvconf entry whenever systemd-resolved itself is stopped?

My current workaround that took me a few hours to figure out is to overwrite the systemd-resolved entries via a network-manager dispatcher script, which due to my very limited testing probably has its own problems.

NetworkManager dispatcher workaround here:
https://gist.github.com/madumlao/68fd4a32e97321e16549287b97f2d1f4

**Comment by floppym** [comment]
This seems to be working just fine with systemd-239. resolved picks up new /etc/hosts entries automatically, and drops entries that have been removed.

Also, a config option was added to disable reading/caching /etc/hosts in https://github.com/systemd/systemd/pull/9742. This has not been included in a release yet.

**Comment by madumlao** [comment]
btw, @floppym, I've noticed that my old hosts entries even persists after reboot for `systemd-resolved`. Would you know where or how I could flush out these hosts entries? They are definitely commented out in /etc/hosts and dnsmasq doesn't pick them up.

**Comment by madumlao** [comment]
Oh I think I get why now. It would appear that systemd does not parse comments in /etc/hosts correctly. Upon completely deleting the comments, systemd-resolved no longer resolves the domain names.

Here's a test so you can reproduce:

1) Create a hosts entry as
~~~
1.2.3.4 google.com www.google.com
~~~
2) dig google.com (you should get 1.2.3.4 as per `/etc/hosts`)
3) dig www.google.com (you should get 1.2.3.4 as per `/etc/hosts`)
4) Comment out www.google.com
~~~
1.2.3.4 google.com # www.google.com
~~~
5) dig www.google.com

Expected behavior: (5) www.google.com should resolve to the public IP.
Actual behavior: (5) www.google.com gives 1.2.3.4

Bonus:

6) dig '#' (you should get 1.2.3.4)

---

As per (5) hosts:
    Fields  of  the  entry  are separated by any number of blanks and/or tab characters. Text from a "#" character until the end of the line is a comment, and is ignored.

It would appear that systemd only treats a line as a comment if the line itself begins with a `#`, although the actual syntax as parsed by most tools allows for a comment at the end of any field. So this is a bug in the parsing of `/etc/hosts` at least.

**Comment by boucman** [comment]
@madumlao it might be a good idea to rename the issue at this point...

**Comment by madumlao** [comment]
Just want to clarify the different issues raised here as there are several.

The initial cause for concern is how systemd-resolved treats `/etc/hosts` entries. Based on the clarifying posts, it appears that systemd does not "cache" `/etc/hosts`, but rather triggers a read of `/etc/hosts`, matching the original request.

The root cause of the report is the bug in reading comments in `/etc/hosts`. I believe this is acknowledged and what has been marked for v240.

A tangent concern that I discovered is how systemd-resolved interacts with resolvconf and dnsmasq. I asked if anyone knows what inserts the `/run/resolvconf/interfaces/systemd-resolved` entry as that breaks some systems when resolved is disabled. I don't yet know what does, thus the NetworkManager workaround I posted. Does anyone know?
